### ISSUE TITLE
Possible binding/route table update inconsistency affects queue operations

### ISSUE THREAD

**Issue by michaelklishin** [rabbitmq/rabbitmq-server#1873]
## Overview

This is for something that's been reported every few weeks on the list for a few years now (e.g. [one example](https://groups.google.com/forum/#!topic/rabbitmq-users/3tlWiaeRBDc) that we never managed to reproduce it, even if others [supposedly can](https://groups.google.com/d/msg/rabbitmq-users/ble31gjFGUE/-rYfRf1JDAAJ). Filing it here so that there is an issue number that's public.

The issue leads to certain operations on a queue failing with a log message that says something like this

```
no binding {queue} between exchange {exchange} in vhost {vhost}
```

and they persist until the queue is deleted and redeclared.

## Current Hypothesis and Findings

One working theory is that it's fundamentally something similar to #1691 in a different place: schema data store transaction retries can affect binding/route table consistency in unexpected ways. That can lead to other parts of the code finding stray or missing bindings they do not expect.

Another is that there's a race condition between node monitor operations on different nodes
which can delete queues and their dependent bindings concurrently in separate transactions.

The log entry is reported by a function that's delegated to by `{queue, exchange}.{bind, unbind}` operations in [rabbit_channel](https://github.com/rabbitmq/rabbitmq-server/blob/ed3dd4d6257df0925ce78ad94de099304346da4c/src/rabbit_channel.erl#L1850). The eventually go through `rabbitmq_binding:add/3` and `rabbit_binding:remove/3` that [perform a `rabbit_route` read](https://github.com/rabbitmq/rabbitmq-server/blob/ed3dd4d6257df0925ce78ad94de099304346da4c/src/rabbit_binding.erl#L165).

## Relevant Changes Merged

#1878, #1884

## Known Ways to Reproduce

#1884 review notes [contain a synthetic way of reproducing](https://github.com/rabbitmq/rabbitmq-server/pull/1884#issuecomment-464277810). If you have a way to reproduce at least semi-reliably
without directly messing with schema tables, **please share it on the mailing list**.

Somewhat relevant here: #1832.

**Comment by michaelklishin** [comment]
#1872 mentions that one way of reproducing is to stop RabbitMQ on all cluster nodes at the same time. While this is not a scenario we find realistic and "solvable", it can be tried to gain improved understanding.

**Comment by michaelklishin** [comment]
Not scheduling for any milestone because we are not confident that #1878/#1879 address every reported case.

**Comment by michaelklishin** [comment]
We introduced 3 different changes that should at least narrow this issue down significantly (in one case I was able to reproduce the report by wiping one of the internal tables from a node's REPL): #1878 (#1879 is a backport to `v3.7.x`, #1884 (will be backported to `v3.7.x` soon) and previously #1721 (#1833 is a backport to `v3.7.x) which has other benefits.

Currently they are all scheduled to ship with `3.7.13`. This issue will remain open for some time, we'd like to see some more real world anecdotal evidence of their effectiveness.
